apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.Tests/V1/PgNumericTests.cs

```diff
@@ -0,0 +1,325 @@
+﻿// Copyright 2022 Google LLC
+// 
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+// 
+//     https://www.apache.org/licenses/LICENSE-2.0
+// 
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using System;
+using System.Collections.Generic;
+using System.Numerics;
+using Xunit;
+
+namespace Google.Cloud.Spanner.V1.Tests
+{
+    public class PgNumericTests
+    {
+        public static IEnumerable<object[]> GetRawValueAndScales()
+        {
+            yield return new object[] { BigInteger.Pow(10, 147455) - 1, 0, true };
+            yield return new object[] { BigInteger.Pow(10, 147455), 0, false };
+            yield return new object[] { -BigInteger.Pow(10, 147455), 0, false };
+            yield return new object[] { -BigInteger.Pow(10, 147455) + 1, 0, true };
+
+            yield return new object[] { BigInteger.Pow(10, 147455) - 1, 110, true };
+            yield return new object[] { BigInteger.Pow(10, 147455) + 1, 110, false };
+
+            yield return new object[] { ((BigInteger.Pow(10, 131072) - 1)/BigInteger.Pow(10, 16383)), 16383, true };
+            yield return new object[] { -((BigInteger.Pow(10, 131072) - 1)/BigInteger.Pow(10, 16383)), 16383, true };
+
+            yield return new object[] { ((BigInteger.Pow(10, 131072) - 1) / BigInteger.Pow(10, 16383)), 16384, false };
+            yield return new object[] { -((BigInteger.Pow(10, 131072) - 1) / BigInteger.Pow(10, 16383)), 16384, false };
+
+            yield return new object[] { BigInteger.Pow(10, 147455) - 1, 16383, true };
+            yield return new object[] { -BigInteger.Pow(10, 147455) + 1, 16383, true };
+        }
+
+        public static IEnumerable<object[]> GetStringInputs()
```

Comment:
Split the data and the tests in two, one for success and one for failure. On the failure test you can use Assert.ThrowsException...

Also, avoid using PgNumeric.ToString() to build the data as we could potentially have the same bug on ToString and Parse. Use the Enumerable approach I descibed above to build the strings.

Reply:
And als add more general type cases, i.e. normal numbers 1 or 2 only, but don't test only the extremes.

Reply:
> Split the data and the tests in two, one for success and one for failure. On the failure test you can use Assert.ThrowsException...
>   There is already a Valid and InValid test for Parse and TryParse that does that.
 
> Also, avoid using PgNumeric.ToString() to build the data as we could potentially have the same bug on ToString and Parse. Use the Enumerable approach I descibed above to build the strings.

PgNumeric.ToString() is just one of the input data. The corresponding input is also provided using BigInteger and have also added Enumerable approach as well as suggested by you.
Also added general test cases.